### PR TITLE
Merging PR #8, PR #9, PR #10, PR #11, PR #12

### PR DESCRIPTION
**Pull request by murtaza98**
This PR attempts to merge the following PR's

1. PR #8 
2. PR #9 
3. PR #10 
4. PR #11 
5. PR #12

### List of all features included
1. Basic RocketChat - Dialogflow interaction
2. API  / Async-hook to `close-chat` and `perform-handover` #6  #3 
3. Automatically perform handover upon multiple fallbacks #5  
4. A caching mechanism for JWT used for Dialogflow authentication #4 
5. Provides a Readme which includes installation, setup, API and App Setting description

### CODE REVIEW

File: AppsDialogflowApp.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            security: ApiSecurity.UNSECURE,
+            endpoints: [
+                new CloseChat(this),
+                new PerformHandover(this),
+            ],
+        });
+        this.getLogger().log('Apps.Dialogflow App Initialized');
+    }
+
+    public async executePostMessageSent(message: IMessage,
```
Comment (renatobecker) [inline AppsDialogflowApp.ts:43]:
```suggestion
    public async executePostMessageSent(message: ILivechatMessage, read: IRead, http: IHttp, persistence: IPersistence, modify: IModify): Promise<void> {
        const handler = new PostMessageSentHandler(this, read, http, modify);
        await handler.run(message);
    }
```

Please, reduce the unnecessary line breaks. Besides that, there is a lot of indentation issues, you can set up your editor to fix this automatically.

File: AppsDialogflowApp.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    public async onSettingUpdated(setting: ISetting, configurationModify: IConfigurationModify, read: IRead, http: IHttp): Promise<void> {
+        const onSettingUpdatedHandler: OnSettingUpdatedHandler = new OnSettingUpdatedHandler(this, read, http);
+        await onSettingUpdatedHandler.run();
+    }
+
+    public async executePostLivechatAgentAssigned(context: ILivechatEventContext, read: IRead, http: IHttp, persistence: IPersistence): Promise<void> {
+        const postLivechatAgentAssignedHandler = new PostLivechatAgentAssignedHandler(context, read, http, persistence);
+        await postLivechatAgentAssignedHandler.run();
```
Comment (renatobecker) [inline AppsDialogflowApp.ts:63]: I don't know why do you need this method, can you give some context about this method?

File: DialogflowWrapper.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
+import { IHttp, IHttpRequest } from '@rocket.chat/apps-engine/definition/accessors';
+import { createSign } from 'crypto';
+import { base64urlEncode } from './helper';
+
+export class DialogflowWrapper {
```
Comment (renatobecker) [inline DialogflowWrapper.ts:5]: Why do you need a class here?
You should create a `helper` folder and implement these helper functions there, you need to export/import the functions you need.

Example:
`helper/Dialogflow.ts`
 
**Dialogflow.ts**
```
export getAccessToken....
...
```

File: app.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+    "id": "783d8e4d-b06a-409a-aaf3-b37650dc0a26",
+    "version": "0.0.1",
+    "requiredApiVersion": "^1.4.0",
+    "iconFile": "icon.png",
```
Comment (renatobecker) [inline app.json:5]: This file is missing.

File: app.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
+{
+    "id": "783d8e4d-b06a-409a-aaf3-b37650dc0a26",
+    "version": "0.0.1",
+    "requiredApiVersion": "^1.4.0",
```
Comment (renatobecker) [inline app.json:4]:
```suggestion
    "requiredApiVersion": "^1.15.0",
```

File: app.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    "iconFile": "icon.png",
+    "author": {
+        "name": "Rocket.Chat",
+        "homepage": "https://rocket.chat/",
+        "support": "https://rocket.chat/"
```
Comment (renatobecker) [inline app.json:9]:
```suggestion
        "support": "support@rocket.chat"
```

File: app.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    },
+    "name": "Apps.Dialogflow",
```
Comment (renatobecker) [inline app.json:11]:
```suggestion
    "name": "Dialogflow",
```

File: app.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    "nameSlug": "appsdialogflow",
```
Comment (renatobecker) [inline app.json:12]:
```suggestion
    "nameSlug": "dialogflow",
```

File: app.json [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    "classFile": "AppsDialogflowApp.ts",
+    "description": "Integration between Rocket.Chat and the Dialogflow Chatbot platform"
+}
```
Comment (renatobecker) [inline app.json:15]: Newline missing at the end of the file.

File: definition/IDialogflowAccessToken.ts [hunk 10]
```diff
@@ -0,0 +1,4 @@
+export interface IDialogflowAccessToken {
+    token: string;
+    expiration: Date;
+}
```
Comment (renatobecker) [inline definition/IDialogflowAccessToken.ts:4]: Newline missing at the end of the file.

File: definition/IParsedDialogflowResponse.ts [hunk 11]
```diff
@@ -0,0 +1,4 @@
+export interface IParsedDialogflowResponse {
+    message: string;
+    isFallback: boolean;
+};
```
Comment (renatobecker) [inline definition/IParsedDialogflowResponse.ts:4]: Newline missing at the end of the file.

File: definition/IParsedDialogflowResponse.ts [hunk 12]
```diff
@@ -0,0 +1,4 @@
+export interface IParsedDialogflowResponse {
```
Comment (renatobecker) [inline definition/IParsedDialogflowResponse.ts:1]:
```suggestion
export interface IDialogflowResponse {
```

File: docs/api-endpoints/close-chat.md [hunk 13]
```diff
@@ -0,0 +1,63 @@
+**Close Chat**
```
Comment (renatobecker) [inline docs/api-endpoints/close-chat.md:1]: We'll remove this endpoint.

File: docs/api-endpoints/perform-handover.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
+**Perform Handover**
```
Comment (renatobecker) [inline docs/api-endpoints/perform-handover.md:1]: We will remove this endpoint.

File: endpoints/CloseChat.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
+import { HttpStatusCode, IHttp, IModify, IPersistence, IRead } from '@rocket.chat/apps-engine/definition/accessors';
```
Comment (renatobecker) [inline endpoints/CloseChat.ts:1]: We'll remove this endpoint.

File: endpoints/PerformHandover.ts [hunk 16]
```diff
@@ -0,0 +1,73 @@
+import { HttpStatusCode, IHttp, IModify, IPersistence, IRead } from '@rocket.chat/apps-engine/definition/accessors';
```
Comment (renatobecker) [inline endpoints/PerformHandover.ts:1]: We'll remove this endpoint.

File: handler/OnSettingUpdatedHandler.ts [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+        if (clientEmail.length === 0 || privateKey.length === 0) {
+            this.app.getLogger().error('Client Email or Private Key Field cannot be empty');
+            return;
+        }
+
+        const dialogflowWrapper: DialogflowWrapper = new DialogflowWrapper(clientEmail, privateKey);
+
+        try {
+            const accessToken = dialogflowWrapper.getAccessToken(this.http);
```
Comment (renatobecker) [inline handler/OnSettingUpdatedHandler.ts:22]: `accessToken` is declared but never used.

File: handler/PostLivechatAgentAssignedHandler.ts [hunk 18]
```diff
@@ -0,0 +1,38 @@
+import { ILivechatEventContext, ILivechatRoom } from '@rocket.chat/apps-engine/definition/livechat';
```
Comment (renatobecker) [inline handler/PostLivechatAgentAssignedHandler.ts:1]: I have no idea why do you need this file.

File: handler/PostMessageSentHandler.ts [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+                private http: IHttp,
+                private persis: IPersistence,
+                private modify: IModify) {}
+
+    public async run() {
+        const SettingBotUsername: string = await getAppSetting(this.read, AppSettingId.DialogflowBotUsername);
+        if (this.message.sender.username === SettingBotUsername) {
+            // this msg was sent by the Bot itself, so no need to respond back
+            return;
+        } else if (this.message.room.type !== RoomType.LIVE_CHAT) {
```
Comment (renatobecker) [inline handler/PostMessageSentHandler.ts:26]: You don't need `else`, you are returning.
```suggestion
        if (this.message.room.type !== RoomType.LIVE_CHAT) {
```

File: handler/PostMessageSentHandler.ts [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+            // check whether this is a Livechat message
+            return;
+        } else if (SettingBotUsername !== getBotUser(this.message).username) {
```
Comment (renatobecker) [inline handler/PostMessageSentHandler.ts:29]: You don't need else, you are returning.
```suggestion
        if (SettingBotUsername !== getBotUser(this.message).username) {
```

File: handler/OnSettingUpdatedHandler.ts [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    public async run() {
+        const clientEmail: string = await getAppSetting(this.read, AppSettingId.DialogflowClientEmail);
+        const privateKey: string = await getAppSetting(this.read, AppSettingId.DialogFlowPrivateKey);
+
+        if (clientEmail.length === 0 || privateKey.length === 0) {
+            this.app.getLogger().error('Client Email or Private Key Field cannot be empty');
+            return;
+        }
+
+        const dialogflowWrapper: DialogflowWrapper = new DialogflowWrapper(clientEmail, privateKey);
```
Comment (renatobecker) [inline handler/OnSettingUpdatedHandler.ts:19]: You don't need to instantiate a new class every time,  you could export the new class in the file you're importing it:
Example:

File **lib/DialogflowWrapper:**

```
class DialogflowWrapper {
...
}

export const DialogflowWrapper = new DialogflowWrapper();
```

File: DialogflowWrapper.ts (hunk 3 above)
Comment (murtaza98) [reply DialogflowWrapper.ts:5]: Sorry, this file shouldn't exist. It should've deleted it during the merge. Kindly ignore this. I'll delete it in the next commit.

File: app.json (hunk 4 above)
Comment (murtaza98) [reply app.json:5]: Could u recheck this? It's present in this PR branch. REF: https://github.com/RocketChat/Apps.Dialogflow/tree/merge-all

File: app.json (hunk 7 above)
Comment (murtaza98) [reply app.json:11]: What do u think of this name -> `Dialogflow Plugin`

File: AppSettings.ts [hunk 22]
```diff
@@ -0,0 +1,64 @@
+import { ISetting, SettingType} from '@rocket.chat/apps-engine/definition/settings';
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply AppSettings.ts]: Renamed it to `Settings.ts`

File: AppSettings.ts [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+    FallbackTargetDepartment = 'fallback_target_department',
+}
+
+export const AppSettings: Array<ISetting> = [
+    {
+        id: AppSettingId.DialogflowBotUsername,
+        public: true,
+        type: SettingType.STRING,
+        packageValue: '',
+        i18nLabel: 'Bot Username',
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply AppSettings.ts]: Added support for i18n translations. REF: https://github.com/RocketChat/Apps.Dialogflow/blob/merge-all/i18n/en.json

File: AppsDialogflowApp.ts [hunk 24]
```diff
@@ -0,0 +1,65 @@
+import {
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply AppsDialogflowApp.ts:1]: Renamed this file

File: definition/IDialogflowAccessToken.ts (hunk 10 above)
Comment (murtaza98) [reply definition/IDialogflowAccessToken.ts:4]: Fixed this.

File: definition/IParsedDialogflowResponse.ts (hunk 12 above)
Comment (murtaza98) [reply definition/IParsedDialogflowResponse.ts:1]: Renamed this interface

File: definition/IParsedDialogflowResponse.ts (hunk 11 above)
Comment (murtaza98) [reply definition/IParsedDialogflowResponse.ts:4]: Fixed this

File: handler/OnSettingUpdatedHandler.ts (hunk 21 above)
Comment (murtaza98) [reply handler/OnSettingUpdatedHandler.ts:19]: Is there a way to pass some values to constructor using this method?

File: handler/OnSettingUpdatedHandler.ts (hunk 17 above)
Comment (murtaza98) [reply handler/OnSettingUpdatedHandler.ts:22]: Removed this unused variable

File: handler/PostLivechatAgentAssignedHandler.ts (hunk 18 above)
Comment (murtaza98) [reply handler/PostLivechatAgentAssignedHandler.ts:1]: Okay, the reason I'm using this event is to save a user's session in persistent storage and to associate/connect it with `visitor-token` which is required later for `handover` endpoint. (PS. The user's session is the same as `room-id`.)

Comment (murtaza98) [reply handler/PostLivechatAgentAssignedHandler.ts:1]: Over here, I'm first checking if the assigned agent is the Bot Agent. If Yes, then I'm saving the session in persistent storage.

Is this the right way to handle sessions? If there r any suggestions then pls let me know.

File: README.md [hunk 25]
```diff
@@ -1,2 +1,70 @@
...
+
+4. (Optional Step) Lastly you can test your Dialogflow Connection by viewing App Logs. To view the logs, goto App Page (`Setting > Apps > Apps.Dialogflow`). There click on menu item (3 vertical dots icon) and then select `View Logs`. There select the **most recent** `onSettingUpdated` title. If you see `------------------ Google Credentials validation Success ----------------` message, then it means your setup is fine. If you don't see this message, then recheck your Dialogflow credentials.
+
+### Apps.Dialogflow's API
+
+The app provides api's to trigger specific actions. The URL for the API can be found on the Apps Page. Currently the app provides 2 API's to trigger following 2 actions
+
+1. **Close Chat**<br/>
+    To close a chat
+    - REST API Documentation for this endpoint can be found [here](https://github.com/RocketChat/Apps.Dialogflow/docs/api-endpoints/close-chat.md)
```
Comment (xiaoyunwu) [inline README.md:66]: these two md files are missing.

File: README.md [hunk 26]
```diff
@@ -1,2 +1,70 @@
...
+You can find all these credentials in a JSON file, which u can get from [here](https://cloud.google.com/dialogflow/docs/quick/setup#sa-create). Under `Create a service account and download the private key file` section there, you will find detailed instruction about how to get the JSON file. After obtaining this JSON file, you can proceed to the next section.
+
+
+### Rocket.Chat Apps Setup
+
+1. First go ahead n create a Bot User. Login as administrator, then goto `Setting > Users`. There create a new Bot User. This new user should have these 2 roles.</br>
+    1. bot
+    2. livechat-agent
+
+2. Then configure the app to automatically assign a livechat-visitor to this bot. To do so, goto `Setting > Livechat > Routing` or `Setting > Omnichannel > Routing`. There enable `Assign new conversations to bot agent` Setting.
```
Comment (xiaoyunwu) [inline README.md:38]: the livechat-agent and livechat-visitor here refer to omnichannel agent/visitor or just livechat widget related?

Comment (murtaza98) [reply README.md:38]: You're right it refers to `omnichannel agent/visitor`

**Pull request merged**